### PR TITLE
Analytics Basic Work

### LINKED ISSUE #69: Basic instructor analytics
Answered vs unanswered
Accuracy
Average rating vs class

### LINKED ISSUE #70: Basic student analytics
Answered vs unanswered vs class
Accuracy vs class
Average rating vs class
Points vs class

### LINKED ISSUE #71: Backend support for basic analytics
Backend support for basic analytics

### LINKED ISSUE #72: Allow instructor to see student analytics
Allow the instructor to view a specific student's analytics

### PR DESCRIPTION
**Pull request by esteeroramy**
fixes #69 , fixes #70 , fixes #71 , fixes #72 
Instructor view:
image
image
Student View:
image

### CODE REVIEW

**Review by ahmedqarmout2 — APPROVED** [review]
Nice work fam 5/7

File: server/db.js [hunk 1]
```diff
@@ -127,17 +133,6 @@ exports.getUserById = function(userId, callback){
     getUserById(userId, callback);
 }
 
-var getUserById = function(userId, callback) {
```
Comment (mehtadarshan1) [inline server/db.js:130]: Why Is this function deleted? You are still calling getUserById(userId, callback); on line 127

File: views/student-analytics.pug [hunk 2]
```diff
@@ -0,0 +1,22 @@
+.row#student-analytics-card
+  if (isAdmin())
```
Comment (mehtadarshan1) [inline views/student-analytics.pug:2]: Why are you referencing admin rights on student-analytics page? This should be separated in analytics.pug.

**Review by mehtadarshan1 — CHANGES_REQUESTED** [review]
Overall looks good, Just a few notes:
1. I do not understand what the donut chart for points mean. Doesn't give a clear message.
2. The bar graph is not useful for showing how many questions you answered vs the class, You can probably just output text in a good Format to showcase .

Just few opinions, Showed it to other users to get a better feedback


Ahmed: for the graphs, if you hover over parts of the graph it will show you what each part means. Also more work coming for your second point. What we have right now is only the initial work. Nice catches 👍

File: views/student-analytics.pug (hunk 2 above)
Comment (esteeroramy) [reply views/student-analytics.pug:2]: The reason for this is that this file is a shared  component between both admin and student. It is not purely for the student.

File: server/db.js (hunk 1 above)
Comment (ahmedqarmout2) [reply server/db.js:130]: Cuz it's a duplicate. It turns out there another function that does the same thing with the same name. Dups are bad :(

File: views/student-analytics.pug (hunk 2 above)
Comment (mehtadarshan1) [reply views/student-analytics.pug:2]: Can we not isolate the two?

Comment (mehtadarshan1) [reply views/student-analytics.pug:2]: Better way is to reference this in the admin analytics page

File: server/db.js (hunk 1 above)
Comment (mehtadarshan1) [reply server/db.js:130]: true

File: views/student-analytics.pug (hunk 2 above)
Comment (esteeroramy) [reply views/student-analytics.pug:2]: Isolation is not needed since this is a shared component. It would add more files and clutter the codebase

Comment (esteeroramy) [reply views/student-analytics.pug:2]: Another point is that if we split it off the code will will have the exact same file with with a different name and without the if statement. This is duplicate code. We will also have to add identical paths in the backend and have a different name for each path, again code duplication.

Comment (Mohammed-JS-Khan) [reply views/student-analytics.pug:2]: @mehtadarshan1 this is the same way it is in question preview where if the prof is viewing a preview of the question, "if preview", then it adds/removes an extra element. I think this is good.

**Review by Mohammed-JS-Khan — APPROVED** [review]
I approve this message.

**Comment by esteeroramy** [comment]
@mehtadarshan1 regarding your second point. This PR is solely intended for the basic analytics of the student vs class. There are 4 more sections that will still be added, one of which is going to display the students's numbers

File: views/student-analytics.pug (hunk 2 above)
Comment (mehtadarshan1) [reply views/student-analytics.pug:2]: Then what is analytics.pug used for? I thought that is for admin

Comment (esteeroramy) [reply views/student-analytics.pug:2]: `analytics.pug` unifies the 2 types of analytics (student and class) using the materialized functionality `cards` and `tabs`. It also imports all the `CSS` and `JS` files and libraries required for both analytics to run, resulting in no code duplication

Comment (mehtadarshan1) [reply views/student-analytics.pug:2]: I was thinking more for analytics security to not mix up admin and students.

Comment (esteeroramy) [reply views/student-analytics.pug:2]: There are no security issues, admin needs to see both sides. Students will only have access to the student analytics as admin analytics is not an option that they can see

Comment (mehtadarshan1) [reply views/student-analytics.pug:2]: sounds good (Y)

**Review by mehtadarshan1 — APPROVED** [review]

File: views/student-analytics.pug (hunk 2 above)
Comment (esteeroramy) [reply views/student-analytics.pug:2]: 🖖👍☺️

**Review by Mohamed-A-Mohamed — APPROVED** [review]
Looks good, nice work

**Pull request merged**